core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -75,6 +79,113 @@ public WriteResult complete() throws IOException {
         .build();
   }
 
+  /**
+   * Base delta writer to write both insert records and equality-deletes.
+   */
+  protected abstract class BaseDeltaWriter implements Closeable {
+    private RollingFileWriter dataWriter;
+    private RollingEqDeleteWriter eqDeleteWriter;
+    private SortedPosDeleteWriter<T> posDeleteWriter;
+    private StructLikeMap<PathOffset> insertedRowMap;
+
+    public BaseDeltaWriter(PartitionKey partition, Schema eqDeleteSchema) {
+      Preconditions.checkNotNull(eqDeleteSchema, "equality-delete schema could not be null.");
+
+      this.dataWriter = new RollingFileWriter(partition);
+
+      this.eqDeleteWriter = new RollingEqDeleteWriter(partition);
+      this.insertedRowMap = StructLikeMap.create(eqDeleteSchema.asStruct());
+
+      this.posDeleteWriter = new SortedPosDeleteWriter<>(appenderFactory, fileFactory, format, partition);
+    }
+
+    /**
+     * Make the generic data could be read as a {@link StructLike}.
+     */
+    protected abstract StructLike asStructLike(T data);
+
+    protected abstract StructLike asCopiedKey(T row);
+
+    public void write(T row) throws IOException {
+      PathOffset pathOffset = PathOffset.of(dataWriter.currentPath(), dataWriter.currentRows());
+
+      StructLike copiedKey = asCopiedKey(row);
+      // Adding a pos-delete to replace the old filePos.
+      PathOffset previous = insertedRowMap.put(copiedKey, pathOffset);
+      if (previous != null) {
+        // TODO attach the previous row if has a positional-delete row schema in appender factory.
+        posDeleteWriter.delete(previous.path, previous.rowOffset, null);
+      }
+
+      dataWriter.write(row);
+    }
+
+    /**
+     * Delete the rows with the given key.
+     *
+     * @param key is the projected values which could be write to eq-delete file directly.
+     */
+    public void delete(T key) throws IOException {
```

Comment:
Why pass a key here instead of a row?

It seems to be that it would be easier to assume that this is a row, so that the `write` and `delete` methods accept the same data.

Reply:
From the tests, I see that this can be used when the deleted row won't be passed. But I don't think this leaves an option for writing an entire row to the delete file and using a subset of its columns for the delete. This assumes that whatever struct is passed here is the entire delete key.

To support both use cases (writing an already projected key and writing a full row and projecting), I think we should have two methods: `deleteKey` and `delete`. That way both are supported. We would also need a test for passing the full row to the delete file, but deleting by key.

Reply:
I agreed that it's better to provide a `delete(T row)` method which can accept an entire row to write eq-delete file. Because it's possible that have a table with (a,b,c,d) columns,  the equality fields are `(a,c)` columns, while someone want to write the `(a,b,c,d)` values into eq-delete file. 

The problem is:  how could we project the generic data type `T` to match the expected `eqDeleteRowSchema` (so that the eqDeleteWriter could write the correct column values) ?   Sounds like we will need to provide an extra abstracted method to accomplish that generic projection ?

Reply:
Thought it again,  if the table has the `(a,b,c,d)` columns, and the equality fields are `(b,c)`,  then I think it's enough to provide two ways to write the equality-delete file: 

1.   Only write the `(b,c)`  column values into equality-delete file.   That is enough to maintain the correct deletion semantics if people don't expect to do any real incremental pulling. 

2. Write the `(a,b,c,d)` column values into equality-delete file.  That's suitable if people want to consume the incremental inserts and deletes for future streaming analysis or data pipeline  etc. 

Except the above cases,  I can't think of other scenarios that require a custom equality schema.

Reply:
Yes, I think those two cases are correct: the caller may want to pass the key to delete, or an entire row to delete.